### PR TITLE
8286941: Add mask IR for partial vector operations for ARM SVE

### PR DESCRIPTION
**Pull request by XiaohongGong**
VectorAPI SVE backend supports vector operations whose vector length is smaller than the max vector length that the current hardware can support. We call them partial vector operations. For some partial operations like vector load/store and the reductions, we need to generate a mask based on the real vector length and use it to control the operations to make sure the results are correct.

For example, if the user defines an IntVector with 256-bit species, and runs it on a SVE hardware that supports 512-bit as the max vector size, all the 256-bit int vector operations are partial. And a mask that all the higher lanes than the real vector length are set to 0 is generated for some ops.

Currently the mask is generated in the backend that is together with the code generation for each op in the match rule. This will generate many duplicate instructions for operations that have the same vector type. Besides, the mask generation is loop invariant which could be hoisted outside of the loop.

Here is an example for vector load and add reduction inside a loop:
```
  ptrue   p0.s, vl8             ; mask generation
  ld1w    {z16.s}, p0/z, [x14]  ; load vector

  ptrue   p0.s, vl8             ; mask generation
  uaddv   d17, p0, z16.s        ; add reduction
  smov    x14, v17.s[0]
```
As we can see the mask generation code "`ptrue`" is duplicated. To improve it, this patch generates the mask IR and adds it to the partial vector ops before code generation. The duplicate mask generation instructions can be optimized out by gvn and hoisted outside of the loop.

Note that for masked vector operations, there is no need to generate additional mask even though the vector length is smaller than the max vector register size, as the original higher input mask bits have been cleared out.

Here is the performance gain for the 256-bit vector reductions work on an SVE 512-bit system:
```
  Benchmark                  size   Gain
  Byte256Vector.ADDLanes     1024   0.999
  Byte256Vector.ANDLanes     1024   1.065
  Byte256Vector.MAXLanes     1024   1.064
  Byte256Vector.MINLanes     1024   1.062
  Byte256Vector.ORLanes      1024   1.072
  Byte256Vector.XORLanes     1024   1.041
  Short256Vector.ADDLanes    1024   1.017
  Short256Vector.ANDLanes    1024   1.044
  Short256Vector.MAXLanes    1024   1.049
  Short256Vector.MINLanes    1024   1.049
  Short256Vector.ORLanes     1024   1.089
  Short256Vector.XORLanes    1024   1.047
  Int256Vector.ADDLanes      1024   1.045
  Int256Vector.ANDLanes      1024   1.078
  Int256Vector.MAXLanes      1024   1.123
  Int256Vector.MINLanes      1024   1.129
  Int256Vector.ORLanes       1024   1.078
  Int256Vector.XORLanes      1024   1.072
  Long256Vector.ADDLanes     1024   1.059
  Long256Vector.ANDLanes     1024   1.101
  Long256Vector.MAXLanes     1024   1.079
  Long256Vector.MINLanes     1024   1.099
  Long256Vector.ORLanes      1024   1.098
  Long256Vector.XORLanes     1024   1.110
  Float256Vector.ADDLanes    1024   1.033
  Float256Vector.MAXLanes    1024   1.156
  Float256Vector.MINLanes    1024   1.151
  Double256Vector.ADDLanes   1024   1.062
  Double256Vector.MAXLanes   1024   1.145
  Double256Vector.MINLanes   1024   1.140
```
This patch also adds 32-bit variants of SVE whileXX instruction with one more matching rule of `VectorMaskGen (ConvI2L src)`. So after this patch, we save one `sxtw` instruction for most VectorMaskGen cases, like below:
```
  sxtw    x14, w14
  whilelo p0.s, xzr, x14  =>  whilelo p0.s, wzr, w14
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8286941](https://bugs.openjdk.org/browse/JDK-8286941): Add mask IR for partial vector operations for ARM SVE

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - Committer) ⚠️ Review applies to [ba59b76e](https://git.openjdk.org/jdk/pull/9037/files/ba59b76e6bb21d4d898e1e3b37e3455edbbfd470)
 * [Ningsheng Jian](https://openjdk.org/census#njian) (@nsjian - Committer) ⚠️ Review applies to [04a2d827](https://git.openjdk.org/jdk/pull/9037/files/04a2d82789f32011524d4658f8c913be7b175ac1)
 * [Nick Gasson](https://openjdk.org/census#ngasson) (@nick-arm - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9037/head:pull/9037` \
`$ git checkout pull/9037`

Update a local copy of the PR: \
`$ git checkout pull/9037` \
`$ git pull https://git.openjdk.org/jdk pull/9037/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9037`

View PR using the GUI difftool: \
`$ git pr show -t 9037`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9037.diff">https://git.openjdk.org/jdk/pull/9037.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Changes I significant. I suggest to wait JDK 20 (next week).

**Comment by XiaohongGong** [comment]
> 

Sure. It's fine for me to wait the JDK 20. Thanks a lot for the advice!

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.cpp:1821]: Due to #8877 such length check may not correct here.
And I don't see `in(2)->Opcode() == Op_VectorMaskGen` check.

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.cpp:1008]: Should this call `StoreVectorNode::Ideal`?

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.cpp:988]: Should this call `LoadVectorNode::Ideal`?
I understand you did optimization because `vector_needs_partial_operations` is false for `LoadVectorMaskedNode` in aarch64 case. But what if it is different on some other (not current) platform?

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.cpp:951]: Why you need `as_LoadVector()` fro `this`? Same in `StoreVectorNode::Ideal().

File: src/hotspot/share/opto/vectornode.cpp
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.cpp:868]: Add assert that only VectorMaskOpNode and ReductionNode expected here.

File: src/hotspot/share/opto/matcher.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/matcher.cpp:2255]: Thanks for looking at this PR! This is removed since we added two combined rules in the aarch64_sve.ad (line-2383) like:
```
match(Set dst (VectorLoadMask (LoadVectorMasked mem pg)));
```
Setting `Op_LoadVectorMasked` as a root in matcher will make such rules not be matched. I'm also curious why this op is  added here. Does it have any influence if I remove it? Thanks!

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:868]: We have other vectornodes like `VectorMaskCmp` , `MaskAll` and `VectorLoadMask` also needs to append the mask here. Actually most masked vector nodes accept the mask input except for the load/store/gather/scatter. And in future, we may extend this to other normal vector nodes whose vector length is full-size while not partial, since SVE always needs a predicate for most instructions. So the default patch will be used for most vector nodes.

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:951]: Good catch and thanks! We could directly use "vect_type()" here. I will change this later.

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:988]: Right, calling `LoadVectorNode::Ideal()` is better. I will change this later. Thanks.

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1008]: ditto

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1821]: I think changes in #8877 influences the max vector length in superword? And since `MaskAll` is used for VectorAPI, the `MaxVectorSize` is still the right reference? @jatin-bhateja, could you please help to check whether this has any influence on x86 avx-512 system? Thanks so much!

> And I don't see in(2)->Opcode() == Op_VectorMaskGen check.

Yes, the `Op_VectorMaskGen` is not generated for `MaskAll` when its input is a constant. We directly transform the `MaskAll` to `VectorMaskGen` here, since they two have the same meanings. Thanks!

File: src/hotspot/share/opto/matcher.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/matcher.cpp:2255]: @jatin-bhateja, could you please help to check the influence about removing this? Kindly know your feedback about this. Thanks so much!

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1821]: >> And I don't see in(2)->Opcode() == Op_VectorMaskGen check.

>Yes, the Op_VectorMaskGen is not generated for MaskAll when its input is a constant. We directly transform the MaskAll to VectorMaskGen here, since they two have the same meanings. Thanks!

I'm sorry that my comment in line-1819 is not right which misunderstood you. I will change this later. Thanks!

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp]: I prefer to not transform `MaskAll` to `VectorMaskGen` now, since there are the match rules using `MaskAll m1` both in sve and avx-512. Doing the transformation may influence those rules.

**Comment by XiaohongGong** [comment]
Hi @vnkozlov ,  all your review comments are resolved. Could you please help to take a look at this PR again? Thanks so much!

**Review by vnkozlov — COMMENTED** [review]
Thank you for addressing my comments.
You need review from Jatin and someone familiar with aarch64 code (I did not look on it).

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:869]: OK, I will add the comment later. Thanks!

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:893]: Added the comment as suggested. Thanks!

**Comment by XiaohongGong** [comment]
> Thank you for addressing my comments. You need review from Jatin and someone familiar with aarch64 code (I did not look on it).

Sure, thanks a lot for the review!

File: src/hotspot/share/opto/matcher.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/matcher.cpp:2255]: Thanks a lot for looking at this part!

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:864]: Yes, I agree with you to add the SuperWordMaxVectorSize reference. Thanks! I will change it.

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1669]: The main reason to add the transformation here is: the FirstTrue needs the reference to the real vector length for SVE, that we need to generate a predicate when the vector length is smaller than the max vector size.  Please check the changes of `partial_op_sve_needed` in aarch64_sve.ad.

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1013]: Actually we don't need to add an additional mask for these masked nodes for SVE. Please refer to the `partial_op_sve_needed` method in `aarch64_sve.ad`, which lists all the ops that needs this transformation. Thanks! `LoadVectorNode::Ideal(phase, can_reshape)` just because `LoadVectorMaskedNode` derived from `LoadVectorNode`, that it may derive its other transformations in future or for other platforms.

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:1033]: Same as `LoadVectorMaskedNode`, `StoreVectorMaskedNode` doesn't need the transformation as well.

File: src/hotspot/share/opto/vectornode.cpp
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:864]: I'm afraid that we cannot directly use `MIN(SuperWordMaxVectorSize, MaxVectorSize) here. As I know `SuperWordMaxVectorSize` is used to control the max vector limit specially for auto-vectorization. It should not have any influence  to the VectorAPI max vector size. So if the supported max vector size for VetorAPI is larger than auto-vectorization, the transformation will be influenced. For example, if a SVE hardware supported 128-bytes max vector size, but the `SuperWordMaxVectorSize` is 64, the predicate will not be generated for vectors whose vector size is smaller than 128-bytes. And I think x86 also has the similar issue. I think we'd better need a unit method to compute the `max_vector_size` that can handle the differences for superword and VectorAPI. And then all the orignal max_vector_size should be replaced with it. WDYT?

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:864]: BTW, the max vector size used here should be referenced from the hardware supported max vector size, which should be `MaxVectorSize` for SVE. For those vectors whose vector size is `SuperWordMaxVectorSize`, but smaller than the hardware supported max size, we still need to generate a predicate for them to make sure the results are right. So using `MaxVectorSize` is necessary here.

Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectornode.cpp:864]: VectorNode::try_to_gen_masked_vector is a common routine which will get called during idealization of all the child vector nodes (generated either by autovectorizer or vectorAPI) unless child overrides its ideal routine.

Will this not result into out of memory writes for storeVector since autovectorizer works under influence of SuperWordMaxVectorSize which could be less than MaxVecotSize but your mask computation is  only considering MaxVectorSize, hence may generate wider masks than desired.

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:864]: The mask is generated based on the real vector length, which guarantees only the vector lanes that inside of its valid vector length will do the operations. Higher lanes do nothing. So if a VectorStore has a vector length which equals to `SuperWordMaxVectorSize`, but the MaxVectorSize is larger than the `SuperWordMaxVectorSize`, a mask will be generated based on the real length `SuperWordMaxVectorSize`. And only lanes under the `SuperWordMaxVectorSize` will be stored. So I cannot see the influence on the out of memory issue. Could you please give an example apart of x86 cases? Thanks so much!  

BTW, I think flag `SuperWordMaxVectorSize`  should be equal to `MaxVectorSize` by default for other platforms instead of 64. If a platform (liken SVE) supports >64 bytes max vector size for auto-vectorization, it can only generate the vectors with 64 bytes vector size. This seems unreasonable. So do you have any plan refactoring this flag, or removing the usages for it in future? Thanks!

Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectornode.cpp:864]: > The mask is generated based on the real vector length, which guarantees only the vector lanes that inside of its valid vector length will do the operations. Higher lanes do nothing. So if a VectorStore has a vector length which equals to `SuperWordMaxVectorSize`, but the MaxVectorSize is larger than the `SuperWordMaxVectorSize`, a mask will be generated based on the real length `SuperWordMaxVectorSize`. And only lanes under the `SuperWordMaxVectorSize` will be stored. So I cannot see the influence on the out of memory issue. Could you please give an example apart of x86 cases? Thanks so much!
> 
Correct, got it 
> BTW, I think flag `SuperWordMaxVectorSize` should be equal to `MaxVectorSize` by default for other platforms instead of 64. If a platform (liken SVE) supports >64 bytes max vector size for auto-vectorization, it can only generate the vectors with 64 bytes vector size. This seems unreasonable. So do you have any plan refactoring this flag, or removing the usages for it in future? Thanks!

File: src/hotspot/cpu/aarch64/assembler_aarch64.hpp
Comment (nsjian) [inline src/hotspot/cpu/aarch64/assembler_aarch64.hpp]: Hi @nsjian, thanks for making a comment in an OpenJDK project!

All comments and discussions in the OpenJDK Community must be made available under the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/). If you already are an OpenJDK [Author](https://openjdk.java.net/bylaws#author), [Committer](https://openjdk.java.net/bylaws#committer) or [Reviewer](https://openjdk.java.net/bylaws#reviewer), please click [here](https://bugs.openjdk.java.net/secure/CreateIssue.jspa?pid=11300&issuetype=1) to open a new issue so that we can record that fact. Please Use "Add GitHub user nsjian" for the summary.

If you are not an OpenJDK Author, Committer or Reviewer, simply check the box below to accept the OpenJDK Terms of Use for your comments.

 - [ ]  I agree to the [OpenJDK Terms of Use](https://openjdk.java.net/legal/tou/) for all comments I make in a project in the [OpenJDK GitHub organization](https://github.com/openjdk).

Your comment will be automatically restored once you have accepted the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/).

Comment (XiaohongGong) [reply src/hotspot/cpu/aarch64/assembler_aarch64.hpp]: Thanks for the comments! I'v updated the patch.

**Review by nsjian — APPROVED** [review]
Hi @nsjian, thanks for making a comment in an OpenJDK project!

All comments and discussions in the OpenJDK Community must be made available under the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/). If you already are an OpenJDK [Author](https://openjdk.java.net/bylaws#author), [Committer](https://openjdk.java.net/bylaws#committer) or [Reviewer](https://openjdk.java.net/bylaws#reviewer), please click [here](https://bugs.openjdk.java.net/secure/CreateIssue.jspa?pid=11300&issuetype=1) to open a new issue so that we can record that fact. Please Use "Add GitHub user nsjian" for the summary.

If you are not an OpenJDK Author, Committer or Reviewer, simply check the box below to accept the OpenJDK Terms of Use for your comments.

 - [ ]  I agree to the [OpenJDK Terms of Use](https://openjdk.java.net/legal/tou/) for all comments I make in a project in the [OpenJDK GitHub organization](https://github.com/openjdk).

Your comment will be automatically restored once you have accepted the OpenJDK [Terms of Use](https://openjdk.java.net/legal/tou/).

**Comment by XiaohongGong** [comment]
> AArch64 changes look good to me.

Thanks for the review @nsjian !

**Comment by vnkozlov** [comment]
I submitted testing. Will let you know results before approval.

**Comment by XiaohongGong** [comment]
> I submitted testing. Will let you know results before approval.

Thanks a lot for doing this!

**Comment by vnkozlov** [comment]
I got failure in tier1:
```
test/hotspot/gtest/aarch64/test_assembler_aarch64.cpp:48: Failure
Expected equality of these values:
  insns[i]
    Which is: 624694305
  insns1[i]
    Which is: 624690209
Ours:
[MachCode]
  0x0000fffca812c698: 2104 3c25 
[/MachCode]
Theirs:
[MachCode]
  0x0000fffc9599d188: 2114 3c25 
[/MachCode]
```

**Comment by vnkozlov** [comment]
I put output into RFE comment.

**Comment by XiaohongGong** [comment]
> I got failure in tier1:
> 
> ```
> test/hotspot/gtest/aarch64/test_assembler_aarch64.cpp:48: Failure
> Expected equality of these values:
>   insns[i]
>     Which is: 624694305
>   insns1[i]
>     Which is: 624690209
> Ours:
> [MachCode]
>   0x0000fffca812c698: 2104 3c25 
> [/MachCode]
> Theirs:
> [MachCode]
>   0x0000fffc9599d188: 2114 3c25 
> [/MachCode]
> ```

Thanks for pointing out this! My fault. I called wrong assembler function in the new added tests.  I will fix it soon!

**Comment by XiaohongGong** [comment]
Hi @vnkozlov , the fix is pushed to this PR. Would you mind running the test again? Thanks so much!

File: src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp
Comment (nick-arm) [inline src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp:1409]: Why not move these two lines into an `else {` block and then you can get rid of the early returns on lines 1398, 1401, and 1404? I think that would make the logic easier to follow.

**Comment by XiaohongGong** [comment]
> An other test failed in tier2: compiler/loopopts/superword/TestPickFirstMemoryState.java Details are in RFE.

Thanks for the tests again! It seems this is the same issue with https://github.com/openjdk/jdk/pull/2867 that the type of `in(2)` is `Type::TOP`. We need to add save the vector type for `ReductionNode`. I will fix it soon!

**Comment by XiaohongGong** [comment]
Hi @vnkozlov , the fixing is pushed to the PR. I tested the failure case on a avx-512 machine, and the failure gone. Could you please help to run the test again? Thanks a lot!

**Comment by vnkozlov** [comment]
I started new testing.

**Review by vnkozlov — APPROVED** [review]
Results are good.

**Comment by XiaohongGong** [comment]
> Results are good.

Thanks a lot for your time again!

**Comment by XiaohongGong** [comment]
@nick-arm , could you please take a look at this change again? Thanks so much!

**Review by nick-arm — APPROVED** [review]

**Comment by XiaohongGong** [comment]
Thanks for the reivew @nick-arm !

File: src/hotspot/share/opto/vectornode.cpp
Comment (eme64) [inline src/hotspot/share/opto/vectornode.cpp:996]: @XiaohongGong Extremely late review :wink: 

Does this not prevent us from doing the `LoadNode::Ideal` optimizations for the cases where `vector_needs_partial_operations` returns true?

See also: https://bugs.openjdk.org/browse/JDK-8371603

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:996]: If `vector_needs_partial_operations` returns true, then the original `LoadVectorNode` is either transformed to a `LoadVectorMaskedNode` or `nullptr`. So it seems `LoadNode::Ideal` is not called if `try_to_gen_masked_vector` returns `nullptr` and some optmizations are missing? That would be an issue.

Comment (eme64) [reply src/hotspot/share/opto/vectornode.cpp:996]: Yes, exactly. I think instead of:
```
return VectorNode::try_to_gen_masked_vector(phase, this, vt);
```
we should do
```
Node* progress = VectorNode::try_to_gen_masked_vector(phase, this, vt);
if (progress != nullptr) { return progress; }
```
That should be correct, right?

Of course the naming here is a bit confusing, and suggests that this may not be correct. Because `vector_needs_partial_operations` would suggest we _always_ need to do partial operations. And so then we would expect that `try_to_gen_masked_vector` would have to _always_ succeed. And so maybe that is why the reviewers did not think that we should continue with `LoadNode::Ideal` if it fails, I suppose? So I think the names should be changed to `maybe_vector_needs_partial_operations` and `transform_to_partial_vector_if_needed`, or similar. What do you think?

Of course I hit an example in [JDK-8371603](https://bugs.openjdk.org/browse/JDK-8371603) where we don't step over MergeMems, which may already be an issue. But I would still like to find some other examples of missing optimizations. Let's see what I can find. Working through QEMU is a bit slow ;)

Comment (eme64) [reply src/hotspot/share/opto/vectornode.cpp:996]: @XiaohongGong Yes, I was able to find a simple reproducer.

```
// java -Xbatch -XX:CompileCommand=compileonly,Test*::test -XX:CompileCommand=printcompilation,Test*::test -XX:+PrintIdeal TestOptimizeLoadVector.java

import jdk.incubator.vector.VectorSpecies;
import jdk.incubator.vector.IntVector;

public class Test1 {

    static final VectorSpecies<Integer> SPECIES =
                IntVector.SPECIES_256;

    static void test(int[] a) {
        // The LOAD below can be optimized away, and be replaced by the value of v1:
        // LoadVectorNode::Ideal calls LoadNode::Ideal, which looks at the memory
        // input and skips and independent stores, finding a store that matches the
        // exact location. And this store stores the value of v1, so we can replace
        // the LOAD, and just use v1 directly. Hence, the example below should have
        // Only a single load, and 3 stores.
        // HOWEVER: if we somehow exit too early in LoadVectorNode::Ideal, we may
        // never reach LoadNode::Ideal and miss the optimization.
        // This happens on aarch64 SVE with 256bits, when we return true for
        // Matcher::vector_needs_partial_operations, but then do nothing when calling
        // VectorNode::try_to_gen_masked_vector. We just return nullptr instantly,
        // rather than trying the other optimizations that LoadNode::Ideal has to
        // offer.
        IntVector v1 = IntVector.fromArray(SPECIES, a, 0 * SPECIES.length());
        v1.intoArray(a, 1 * SPECIES.length()); // STORE of v1
        v1.intoArray(a, 2 * SPECIES.length()); // independent STORE - no overlap with STORE above and LOAD below.
        IntVector v2 = IntVector.fromArray(SPECIES, a, 1 * SPECIES.length()); // LOAD - is it replaced with v1?
        v2.intoArray(a, 3 * SPECIES.length());
    }

    public static void main(String[] args) {
        int[] a = new int[1000];
        for (int i = 0; i < 10_000; i++) {
            test(a);
	}
    }
}
```

I'll see if we can do similar things for the other cases.

Comment (eme64) [reply src/hotspot/share/opto/vectornode.cpp:996]: We can continue the conversation in:
https://bugs.openjdk.org/browse/JDK-8371603

Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.cpp:996]: > Yes, exactly. I think instead of:
> 
> ```
> return VectorNode::try_to_gen_masked_vector(phase, this, vt);
> ```
> 
> we should do
> 
> ```
> Node* progress = VectorNode::try_to_gen_masked_vector(phase, this, vt);
> if (progress != nullptr) { return progress; }
> ```
> 
> That should be correct, right?

Yes, I think so.

> Of course the naming here is a bit confusing, and suggests that this may not be correct. Because `vector_needs_partial_operations` would suggest we _always_ need to do partial operations. And so then we would expect that `try_to_gen_masked_vector` would have to _always_ succeed. And so maybe that is why the reviewers did not think that we should continue with `LoadNode::Ideal` if it fails, I suppose? So I think the names should be changed to `maybe_vector_needs_partial_operations` and `transform_to_partial_vector_if_needed`, or similar. What do you think?

`vector_needs_partial_operations` is used to check whether we need consider the partial vector lanes for the specified op. The partial vector lanes means the vector size of the op is lower than the MaxVectorSize, hence the higher lanes should be ignored. 

I agree with your point.  I'v created a JBS (https://bugs.openjdk.org/browse/JDK-8372717) to fix it and will create a patch soon. Thank you so much for the debugging, testing and any input!

Comment (eme64) [reply src/hotspot/share/opto/vectornode.cpp:996]: @XiaohongGong I already have 2 reproducers for missing load and store optimizations attached to [JDK-8371603](https://bugs.openjdk.org/browse/JDK-8371603).

You should add those 2 tests with IR verification to your PR then. And you should also verify that this patch fixes the issue from [JDK-8371603](https://bugs.openjdk.org/browse/JDK-8371603), and then we can close that one as a duplicate. It is a JDK26 bug, so we should try to fix it soon-ish ;)

Comment (eme64) [reply src/hotspot/share/opto/vectornode.cpp:996]: You filed [JDK-8372717](https://bugs.openjdk.org/browse/JDK-8372717) as a JDK27 issue. But it really is also a fix for the JDK26 regression of [JDK-8371603](https://bugs.openjdk.org/browse/JDK-8371603). Not sure if we should convert it to a bug, or just tag the PR with both issues, so both are resolved as a consequence of it. The JDK26 fork is soon, and so if we don't integrate before that we would have to backport, I think.

**Pull request closed without merging**
